pkg/ring/kv/etcd/etcd.go

```diff
@@ -129,13 +129,19 @@ func (c *Client) WatchKey(ctx context.Context, key string, f func(interface{}) b
 		MinBackoff: 1 * time.Second,
 		MaxBackoff: 1 * time.Minute,
 	})
+
+	// Ensure the context used by the Watch is always cancelled.
+	watchCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
Why do we need this?

Reply:
The callback can return false causing us to exit the loop - we need to ensure the context used by the Watch is cancelled in this case.